packages/cursorless-vscode/src/CommandHistoryAnalyzer.ts

```diff
@@ -0,0 +1,141 @@
+  getPartialTargetDescriptors,
+  getScopeType,
+} from "@cursorless/cursorless-engine";
+import globRaw from "glob";
+import { readFile } from "node:fs/promises";
+import path from "node:path";
+import { promisify } from "node:util";
+import * as vscode from "vscode";
+
+const glob = promisify(globRaw);
+
+class Month {
+  readonly month: string;
+  readonly actions: Record<string, number> = {};
+  readonly modifiers: Record<string, number> = {};
+  readonly scopeTypes: Record<string, number> = {};
+  count: number = 0;
+
+  constructor(month: string) {
+    this.month = month;
+  }
+
+  toString(): string {
+    return [
+      `[${this.month}]`,
+      `Total commands: ${this.count}`,
+      this.serializeMap("Actions", this.actions),
+      this.serializeMap("Modifiers", this.modifiers),
+      this.serializeMap("Scope types", this.scopeTypes),
+    ].join("\n\n");
+  }
+
+  private serializeMap(name: string, map: Record<string, number>) {
+    const entries = Object.entries(map);
+    entries.sort((a, b) => b[1] - a[1]);
+    const entriesSerialized = entries
+      .map(([key, value]) => `  ${key}: ${value}`)
+      .join("\n");
+    return `${name} (${entries.length}):\n${entriesSerialized}`;
+  }
+
+  append(entry: CommandHistoryEntry) {
+    this.count++;
+    this.incrementAction(entry.command.action.name);
+
+    const partialTargets = getPartialTargetDescriptors(entry.command.action);
+    const partialPrimitiveTargets = getPartialPrimitiveTargets(partialTargets);
+    this.parsePrimitiveTargets(partialPrimitiveTargets);
+  }
+
+  private parsePrimitiveTargets(
+    partialPrimitiveTargets: PartialPrimitiveTargetDescriptor[],
+  ) {
+    for (const target of partialPrimitiveTargets) {
+      if (target.modifiers == null) {
+        continue;
+      }
+      for (const modifier of target.modifiers) {
+        this.incrementModifier(modifier);
+
+        const scopeType = getScopeType(modifier);
+        if (scopeType != null) {
+          this.incrementScope(scopeType);
+        }
+      }
+    }
+  }
+
+  private incrementAction(actionName: string) {
+    this.actions[actionName] = (this.actions[actionName] ?? 0) + 1;
+  }
+
+  private incrementModifier(modifier: Modifier) {
+    this.modifiers[modifier.type] = (this.modifiers[modifier.type] ?? 0) + 1;
+  }
+
+  private incrementScope(scopeType: ScopeType) {
+    this.scopeTypes[scopeType.type] =
+      (this.scopeTypes[scopeType.type] ?? 0) + 1;
+  }
+}
+
+class Months {
+  readonly months: Record<string, Month> = {};
+
+  getMonth(entry: CommandHistoryEntry) {
+    const monthDate = entry.date.slice(0, 7);
+
+    if (!this.months[monthDate]) {
+      this.months[monthDate] = new Month(monthDate);
+    }
+
+    return this.months[monthDate];
+  }
+
+  getMonths() {
+    const months = Object.values(this.months);
+    months.sort((a, b) => a.month.localeCompare(b.month));
+    return months;
+  }
+}
+
+async function analyzeFilesAndReturnMonths(dir: string): Promise<Month[]> {
```

Comment:
Why are all your abstractions in terms of months? Feels a bit arbitrary. The roll-up window should just be some knob we can tweak, rather than being deeply baked into the analysis code.

Shall we factor this analysis code into a separate PR so we can review that separately without blocking this one? I'd love to get the history part merged in

Reply:
Just an arbitrary choice that I felt suited it. Easy to change if we want to.
There's no reason why we can't change the analysis code later?

Reply:
I guess it's very self-contained, but I just don't like landing code I haven't reviewed, and I'd love to get this thing in

Reply:
I've gonna have to take the time to move everything to the engine so you have time to look at this if you want to.

Reply:
Atomic PRs

Reply:
Atomic PRs